server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java

```diff
@@ -782,6 +783,27 @@ public final void writeOptionalInstant(@Nullable Instant instant) throws IOExcep
             o.writeInt(period.getYears());
             o.writeInt(period.getMonths());
             o.writeInt(period.getDays());
+        }),
+        entry(GenericNamedWriteable.class, (o, v) -> {
+            final var genericNamedWriteable = (GenericNamedWriteable) v;
+            final var minSupportedVersion = TransportVersion.max(
+                TransportVersion.V_8_500_069,
+                genericNamedWriteable.getMinimalSupportedVersion()
+            );
+
+            if (o.getTransportVersion().before(minSupportedVersion)) {
+                final var message = Strings.format(
+                    "[%s] requires minimal transport version [%s] and cannot be sent using transport version [%s]",
+                    genericNamedWriteable,
+                    minSupportedVersion,
+                    o.getTransportVersion()
+                );
+                assert false : message;
```

Comment:
These asserts seem to cause integration tests to fail in more opaque ways, with the test code receiving less information about the failure. Is this generally a good idea? Is the advantage that we fail faster, and so avoid being mislead by later errors?

Reply:
We definitely want to fail here in tests, and simply throwing an exception won't necessarily do that because exceptions can be caught. It's unfortunate that we have less context to share here, although I would expect these things to be fairly easy to reproduce.

Reply:
At least here we assert with the same message. The test failure I debugged recently failed without the message, hiding the reason for failure, making it much harder to track initially.